enhancements/node-healthcheck-operator.md

```diff
@@ -0,0 +1,435 @@
+Node Health Check Enhancement 
+
+---
+title: node-healthcheck-controller
+authors:
+  - "@rgolangh"
+reviewers:
+  - "@abeekof"
+  - "@n1r1"
+  - "@mshitrit"
+  - "@slintes"
+  - "@dhellmann"
+approvers:
+  - "@abeekof"
+  - "@n1r1"
+---
+
+# Node health checking
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs
+
+
+## Summary
+
+
+Enable opt in for automated node health checking and create objects
+for unhealthy nodes, to be handled later by external controllers
+that comply with the existing  [api].
+
+## Motivation
+
+Enable clusters that do not have a functioning Machine API to detect and respond to node-level failures in order to recover workloads , and/or restore cluster capacity.
+
+The existing Machine Health Check controller requires a functioning machine-api. This prevents its use in the majority of Kubernetes and OpenShift clusters.
```

Comment:
It would be good to have more justification here for why we need that feature for user-managed deployments and why we don't simply recommend creating those deployments with all of the automation that the IPI workflow brings with it.

Reply:
Wouldn't that basically be a justification of why UPI exists?
Given that a majority of deployments are UPI, I'd have thought that to be assumed knowledge.

Reply:
```suggestion
The majority of OpenShift clusters are UPI based and do not have Machine objects
and/or a functioning Machine API - which is a prerequisite for using the Machine 
HealthCheck controller.

There are many possible barriers to retrofitting a Machine API to an existing UPI 
cluster, but the most common raised are: complexity, technical limitations, and 
customer policies which prevent BMCs or platform credentials from being stored 
within the cluster).

While there remains a need for UPI, there is also a need to allow UPI workloads to 
be made highly available.
```

Reply:
I would expect a UPI deployment to use external tools for monitoring, for example, just as it does for provisioning. Similarly, I would expect a user who wants automation for managing their cluster to be fully integrated to use the IPI tools, even if that means redeploying. So, having some sort of detail about cases where the user wants or needs a UPI deployment but also wants the automation described here would help explain why we want to support a new component duplicating the functionality of one we already have.

Reply:
TBH, that seems punitive.  You didn't want X so you can't have Y either.

IIUC there are a variety of reasons, even on cloud, that customers might need to pursue a UPI strategy, but there are zero technical reasons that needs to prevent them from being able to use a standard solution for protecting their workloads.

We both heard customer0 very clearly say they would much rather use a solution from Red Hat than write their own and I see no reason they would be an exception.

I thought we were trying to get away from this idea that how you installed a cluster locks you into a certain feature set.

Reply:
Those are good justifications for solving the problem, and would be good additions to the doc.

Reply:
Answering the easy question first...

> Are we explicitly trying to support platforms that don't currently have Machine API support at all?

@JoelSpeed absolutely. that is the primary goal

> There's nothing that prevents a customer from using Machine API on a UPI cluster, so I don't think we should be implying MHC doesn't work on most clusters. Granted most UPI clusters don't use MAPI, but that's not what is stated here.

I'm not sure I see the distinction...
There may not be an internal technical reason[1], but presumably the majority of customers that deploy with UPI aren't just being lazy and there is something about the customer environment (technical or political) that prevents them turning on MAPI after installation too.

[1] Though based on experience with Assisted Installer and metal clusters, I would argue it is far from a trivial exercise.